### ISSUE TITLE
Moving map sometimes doesn't repaint correctly

### ISSUE THREAD

**Issue by stefanbeller** [mapeditor/tiled#174]
(Moving https://sourceforge.net/apps/mantisbt/tiled/view.php?id=84 to the new issue tracker here)

When the map is being dragged around by either middleclick or by space bar and left mouseclick, sometimes 
artifacts (the current stampbrush) is painted on the map and not removed correctly.

That happens when you scroll against a border (so mouse moves, but map does not move, because map border is reached)
and once you are done scrolling, the area is not repainted until you move the mouse.

**Comment by cjke** [comment]
@bjorn I am still experiencing this issue on build 0.14.2. I'm on a Mac running El Capitan. I have included a few screenshots. Oddly, I can't turn on OpenGL mode. 

I'm not sure of the Qt version, it doesn't seem to be an available option in Help anymore. This is on a fresh install of Tiled. 

![screenshot 2015-10-23 23 33 07](https://cloud.githubusercontent.com/assets/884065/10695506/f566ee52-79de-11e5-8827-6ee3fbfd0c02.png)
![screenshot 2015-10-23 23 33 30](https://cloud.githubusercontent.com/assets/884065/10695507/f5705780-79de-11e5-9c22-452d1a25ca6c.png)

**Comment by bjorn** [comment]
> @bjorn I am still experiencing this issue on build 0.14.2. I'm on a Mac running El Capitan. I have included a few screenshots. Oddly, I can't turn on OpenGL mode.

Yeah, unfortunately this is almost certainly a bug on the Qt side. A simple example application demonstrating the issue may help to get it fixed, but unfortunately this part of Qt has been rather unmaintained in the past years. The issue did get a lot less severe than the initial bug seen in the screenshots posted by @parasyte, right?

You can't turn on OpenGL because I disabled it on OS X. There is no version of Qt 5 where OpenGL has worked in any usable way (see for example issue #529).

> I'm not sure of the Qt version, it doesn't seem to be an available option in Help anymore. This is on a fresh install of Tiled.

Right, I thought it was not important enough to have its own menu item, but I should probably at least put the Qt version into the About dialog. In any case I'm sure it's Qt 5.5.0.

**Comment by cjke** [comment]
Thanks @bjorn for the quick reply. 

The severity seems on par with that shown by @parasyte I would say. 

Makes sense to have it disabled, and makes equal sense for Qt to not have a menu anymore - I was only wanting to include it to make sure my issue was complete. 

It's a shame that it's a bug with the underlying framework, but it's tolerable. Thanks again, and amazing product! Goodluck with the patreon support.

**Comment by bjorn** [comment]
I've created a small example application that reproduces this issue, which confirms it's an issue in Qt:

```c++
#include <QApplication>
#include <QGraphicsRectItem>
#include <QGraphicsScene>
#include <QGraphicsSceneMouseEvent>
#include <QGraphicsView>
#include <cmath>

class Scene : public QGraphicsScene
{
public:
    /**
     * Sets up the scene with a "canvas" and a "cursor".
     */
    Scene()
    {
        addRect(QRectF(0, 0, 1024, 1024));
        cursor = addRect(QRectF(0, 0, 32, 32), QPen(), Qt::lightGray);
    }

protected:
    /**
     * Changes the position of the cursor to stay below the mouse, snapped to a
     * virtual grid.
     *
     * Changing the position of an item in response to a mouse event is
     * normally not a problem, but it causes visible repainting issues when
     * that mouse move event was generated due to handling a mouse wheel event.
     *
     * The snapping is necessary in triggering the issue, because otherwise the
     * item position on the screen would not change at all while scrolling.
     */
    void mouseMoveEvent(QGraphicsSceneMouseEvent *mouseEvent) override
    {
        auto snapped = mouseEvent->scenePos() / 32;
        snapped.setX(std::floor(snapped.x()));
        snapped.setY(std::floor(snapped.y()));
        cursor->setPos(snapped * 32);

        QGraphicsScene::mouseMoveEvent(mouseEvent);
    }

private:
    QGraphicsRectItem *cursor;
};

int main(int argc, char *argv[])
{
    QApplication a(argc, argv);

    Scene scene;

    QGraphicsView view;
    view.setScene(&scene);
    view.viewport()->setMouseTracking(true);
    view.resize(256, 256);  // make sure the view can be scrolled with wheel
    view.show();

    return a.exec();
}
```

Zipped up together with a project file: [repaintbug.zip](https://github.com/mapeditor/tiled/files/6682516/repaintbug.zip)
